kong/clustering.lua

```diff
@@ -0,0 +1,409 @@
+
+  local entities, _, _, _, new_hash = dc:parse_table(config_table)
+  if not entities then
+    return nil, "bad config received from control plane"
+  end
+
+  if declarative.get_current_hash() ~= new_hash then
+    -- NOTE: no worker mutex needed as this code can only be
+    -- executed by worker 0
+    local res, err = declarative.load_into_cache_with_events(entities, new_hash)
+    if not res then
+      return nil, err
+    end
+
+    if update_cache then
+      -- local persistence only after load finishes without error
+      local f, err = io_open(CONFIG_CACHE, "w")
+      if not f then
+        ngx_log(ngx_ERR, "unable to open cache file: ", err)
+
+      else
+        local res
+        res, err = f:write(cjson_encode(config_table))
+        if not res then
+          ngx_log(ngx_ERR, "unable to write cache file: ", err)
+        end
+
+        f:close()
+      end
+    end
+
+  else -- get_current_hash() == new_hash
+    ngx_log(ngx_DEBUG, "same config received from control plane,",
+            "no need to reload")
+    return true
+  end
+
+  return true
+end
+
+
+local function send_ping(c)
+  local _, err = c:send_ping(declarative.get_current_hash())
+  if err then
+    ngx_log(ngx_ERR, "unable to ping control plane node: ", err)
+    -- return and let the main thread handle the error
+    return
+  end
+
+  ngx_log(ngx_DEBUG, "sent PING packet to control plane")
+end
+
+
+local function communicate(premature, conf)
+  if premature then
+    -- worker wants to exit
+    return
+  end
+
+  -- TODO: pick one random CP
+  local address = conf.cluster_control_plane
+
+  local c = assert(ws_client:new(WS_OPTS))
+  local uri = "wss://" .. address .. "/v1/outlet?node_id=" ..
+              kong.node.get_id() .. "&node_hostname=" .. utils.get_hostname()
+  local res, err = c:connect(uri, { ssl_verify = true,
+                                    client_cert = CERT,
+                                    client_priv_key = CERT_KEY,
+                                    server_name = "kong_clustering",
+                                  }
+                            )
+  if not res then
+    local delay = 9 + math.random()
+
+    ngx_log(ngx_ERR, "connection to control plane broken: ", err,
+            " retrying after ", delay , " seconds")
+    assert(ngx.timer.at(delay, communicate, conf))
+    return
+  end
+
+  -- connection established
+  -- ping thread
+  ngx.thread.spawn(function()
+    while true do
+      send_ping(c)
+      ngx_sleep(PING_INTERVAL)
+    end
+  end)
+
+  while true do
+    local data, typ, err = c:recv_frame()
+    if err then
+      ngx.log(ngx.ERR, "error while receiving frame from control plane: ", err)
+      c:close()
+
+      local delay = 9 + math.random()
+      assert(ngx.timer.at(delay, communicate, conf))
+      return
+    end
+
+    if typ == "binary" then
+      local msg = assert(cjson_decode(data))
+
+      if msg.type == "reconfigure" then
+        local config_table = assert(msg.config_table)
```

Comment:
Same thing here.

Reply:
`config_table` missing means the CP sent a malformed message, which is a bug.

Reply:
@dndx the question is, should the thing crash because of that or continue working? The message may get corrupted, timeout'd etc.?

Reply:
Corrupted message means a bug in the code, in case of timeouts the code will never reach here as JSON decoding would not even happen in the first place.